contracts/modules/utils/RequireUtils.sol

```diff
@@ -68,6 +71,51 @@ contract RequireUtils {
     emit RequiredConfig(_wallet, imageHash, _threshold, abi.encode(_members));
   }
 
+  function requireAndIndexConfig(
+    address _wallet,
+    uint256 _threshold,
+    Member[] calldata _members
+  ) external {
+    // Compute expected imageHash
+    bytes32 imageHash = bytes32(uint256(_threshold));
+    for (uint256 i = 0; i < _members.length; i++) {
+      imageHash = keccak256(abi.encode(imageHash, _members[i].weight, _members[i].signer));
+    }
+
+    // Check against wallet imageHash
+    (bool succeed, bytes memory data) = _wallet.call(abi.encodePacked(IModuleAuthUpgradable(_wallet).imageHash.selector));
+    if (succeed && data.length == 32) {
+      // Check contract defined
+      bytes32 currentImageHash = abi.decode(data, (bytes32));
+      require(currentImageHash == imageHash, "RequireUtils#requireAndIndexConfig: UNEXPECTED_IMAGE_HASH");
+    } else {
+      // Check counter-factual
+      require(address(
+        uint256(
+          keccak256(
+            abi.encodePacked(
+              byte(0xff),
+              FACTORY,
+              imageHash,
+              INIT_CODE_HASH
+            )
+          )
+        )
+      ) == _wallet, "RequireUtils#requireAndIndexConfig: UNEXPECTED_COUNTERFACTUAL_IMAGE_HASH");
+
+      // Store counter factual imageHash
+      initialImageHash[_wallet] = imageHash;
+    }
+
+    // Emit event for easy config retrieval
+    emit RequiredConfig(_wallet, imageHash, _threshold, abi.encode(_members));
+
+    // Store block height for first log
+    if (imageHashBlockHeight[imageHash] == 0) {
+      imageHashBlockHeight[imageHash] = block.number;
```

Comment:
Shouldn't we store the latest blockheight as well? Could be millions of blocks between the first one and the most recent update.

Reply:
@PhABC yes agreed. I thought by the design we discussed, this variable would hold the highest block height of the latest image hash update. I notice we also store the initial Image Hash, either that value alone is sufficient for our purposes, or we replace that with initialImageBlockHeight uint256, and we can reconstruct it from the logs

Reply:
The process of retrieving a configuration requires two steps:

1) Knowing the imageHash of the wallet
2) Knowing the configuration of the imageHash

For 1) there are two possibilities:

a) the wallet is in counter-factual state or
b) the wallet has been updated

For a) we use the `initialImageHash` mapping, but once the wallet it's updated and it's no longer counter-factual, the imageHash is stored on the contract storage and we can query it just by calling the contract.

And for 2), each imageHash is just a hash of a configuration, so once it's "map" from hash to configuration is created, there is no need to do further updates, either the wallet uses the imageHash and we already have it mapped, or the imageHash changes and we just add another entry to `imageHashBlockHeight`.